### ISSUE TITLE
PVS-Check: incorrect check in Orleans.UniqueKey.NewKey method

### ISSUE THREAD

**Issue by centur** [dotnet/orleans#1397]
Actually comment is correct and describes different behavior than an actual check

**V3022 Expression 'n1 == 0 && n1 != 0' is always false. Unsigned type value is always >= 0. Probably the '||' operator should be used here. Orleans UniqueKey.cs 113**

```
        private static UniqueKey NewKey(ulong n0, ulong n1, Category category, long typeData, string keyExt)
        {
            // in the string representation of a key, we grab the least significant half of n1.
            // therefore, if n0 is non-zero and n1 is 0, then the string representation will always be
            // 0x0 and not useful for identification of the grain.

            if (n1 == 0 && n1 != 0)
                throw new ArgumentException("n0 cannot be zero unless n1 is non-zero.", "n0");
```

Should be `if (n1 == 0 && n0 != 0)`

**Comment by jthelin** [comment]
Seems like there are slightly conflicting statements here:

in the comment "if n0 is non-zero and n1 is 0" ... then bad things happen ...
  -> `if (n0 != 0 && n1 == 0)` then bad

vs the exception message "n0 cannot be zero unless n1 is non-zero" 
  -> `if (n0 == 0 && n1 != 0) then throw ...`

Do you think we should include which one or both checks @gabikliot @sergeybykov ?

But then how does GrainId == `0L` get represented correctly?

Should the [Grain] `Category == Category.KeyExtGrain` be factored into the comparison because that is what the comment is about?

**Comment by gabikliot** [comment]
I have no idea if that comment is correct or not. I never liked bit tweaking there.

**Comment by sergeybykov** [comment]
Resolved via #2728.
